apollo-runtime/src/main/java/com/apollographql/apollo/internal/RealApolloCall.java

```diff
@@ -191,10 +198,13 @@ private RealApolloCall(Builder<T> builder) {
   }
 
   @Override public void cancel() {
-    canceled = true;
-    interceptorChain.dispose();
-    if (queryReFetcher.isPresent()) {
-      queryReFetcher.get().cancel();
+    synchronized (this) {
```

Comment:
I feel like we need the `canceled = true` and `callback = Optional.absent` to happen atomically. Otherwise we could have a state where a user gets a callback, during which she check `call.isCanceled` and sees that evaluate to `true`... probably not too harmful, but it does seem to break our contract.